src/vsr/replica.zig

```diff
@@ -8968,13 +8981,21 @@ const DVCQuorum = struct {
                 }
 
                 if (header_nacks.isSet(header_index)) {
+                    assert(header.op > dvc.header.checkpoint_op);
+
                     // The op is nacked explicitly.
                     nacks += 1;
-                } else if (vsr.Headers.dvc_header_type(header) == .valid and
-                    header_canonical != null and header_canonical.?.checksum != header.checksum)
-                {
-                    // The op is nacked implicitly, because the replica has a different header.
-                    nacks += 1;
+                } else if (vsr.Headers.dvc_header_type(header) == .valid) {
+                    if (header_canonical != null and header_canonical.?.checksum != header.checksum) {
+                        // The op is nacked implicitly, because the replica has a different header.
+                        nacks += 1;
+                    }
+                    if (header_canonical == null and header.view < log_view_canonical) {
+                        assert(dvc.header.log_view < log_view_canonical);
```

Comment:
This assert can be copied to the `if (header_canonical != null and header_canonical.?.checksum != header.checksum) {` branch too -- if the DVC has an uncanonical header, then it must be an uncanonical DVC.

Reply:
Right! And, taking a second look here, `and header.view < log_view_canonical` is also an assert, rather than a condition! if we don't have a canonical header, then the header we have must be from before!